node/src/components/small_network/tasks.rs

```diff
@@ -429,6 +430,14 @@ where
             return Err(ConnectionError::WrongNetwork(network_name));
         }
 
+        // We check the chainspec hash to ensure peer is running on the
+        // same chainspec as us.
+        if let Some(hash) = chainspec_hash {
```

Comment:
I think I'd prefer to have it non-optional in the `Handshake` variant.

We can still use `#[serde(default)]` with it.

Reply:
Also, we need to move this check to after the protocol version check just below, otherwise the tarpitting will be effectively dead code.

Reply:
We had it be an `Option` because it makes more sense if you want to see if it was included in the handshake message, as a hash itself wouldn't have a default value. It must be included in the message anyway from this version onwards, please see the updated handling of this case.

Reply:
But in the case where you don't get a chainspec hash in the handshake message, and the deserializer fills in the default, `WrongChainspecHash` is not the semantically correct error here because none was actually provided. I think using `Digest::default` actually subtracts value (from a user point of view) and is actually not at all expensive to maintain for us.

Reply:
I'm not a fan of a default `Digest` - if it's not there, we just don't want to use it?